### ISSUE TITLE
Returning string from read includes the terminating \0

### ISSUE THREAD

**Issue by DirtYiCE** [lwoggardner/ffi-libfuse#34]
I'm trying to create fuse filesystem that uses direct_io to create virtual files (like in /proc, has 0 size, but you can read them). Unfortunately, when returning a string, ffi-libfuse uses `FFI::MemoryPointer.from_string` to create a pointer from the string, and such ffi pointers include the C string's terminating `\0` in length.
https://github.com/lwoggardner/ffi-libfuse/blob/1af4e23a87ec48f1390f6a8a9f95eaaf127dd12f/lib/ffi/libfuse/fuse_buf.rb#L89

Here's the hello world fs from the readme, modified to use direct_io on the file:
```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'ffi/libfuse'

# Hello World!
class HelloFS
  include FFI::Libfuse::Adapter::Ruby
  include FFI::Libfuse::Adapter::Fuse2Compat

  # FUSE Configuration methods

  def fuse_options(args)
    args.parse!({ 'subject=' => :subject }) do |key:, value:, **|
      raise FFI::Libfuse::Error, 'subject option must be at least 2 characters' unless value.size >= 2

      @subject = value if key == :subject
      :handled
    end
  end

  def fuse_help
    '-o subject=<subject>   a target to say hello to'
  end

  def fuse_configure
    @subject ||= 'World!'
    @content = "Hello #{@subject}\n"
  end

  # FUSE callbacks

  def getattr(path, stat, *_args)
    case path
    when '/'
      stat.directory(mode: 0o550)
    when '/hello.txt'
      stat.file(mode: 0o440, size: 0)
    else
      raise Errno::ENOENT
    end
  end

  def readdir(_path, *_args)
    yield 'hello.txt'
  end

  def read_buf path, size, offs, ffi
    return StringIO.new unless offs == 0
    StringIO.new @content
  end

  def open path, info
    info.nonseekable = true
    info.direct_io = true
    nil
  end
end

# Start the file system
exit(FFI::Libfuse.fuse_main(operations: HelloFS.new)) if __FILE__ == $0
```

`cat`-ing the output with `-v` shows the `\0` added after `\n`
```shell
$ cat -v hello.txt
Hello World!
^@
```

(Also note that `StringIO.new.read(some_number)` returns `nil` and that doesn't cause any more bytes to be returned by `read`, simply doing `return '' unless offs == 0` results in an infinite loop with `cat`, as every read returns a single zero byte.)

**Comment by lwoggardner** [comment]
Thanks for the report.

Notes/Thoughts  as I try to work out where to start with this.

If not from_string (can see the implementation explicitly adds +1 to len),  should it be ~~put_string~~ or ~~write_string~~, or **write_bytes**

~~Does FuseBufVec.create , sending str: nil  to the intiial fuse buf on eof (nil read from IO) the right behaviour~~
https://github.com/lwoggardner/ffi-libfuse/blob/1af4e23a87ec48f1390f6a8a9f95eaaf127dd12f/lib/ffi/libfuse/fuse_buf_vec.rb#L59
No - an explicitly nil str is protected and results in an empty FuseBuf,  but non empty is being handled with from_string

Can this direct IO scenario be handled without adapter Ruby, adapter Fuse2Compat - just to keep things simple for testing.

**Comment by lwoggardner** [comment]
Have replicated

Fix is to allocate the FFI::MemoryPointer and push contents with write_bytes.  Hope to release tomorrow.

**Comment by DirtYiCE** [comment]
Checked the newly released 0.4.4, works fine now, thanks! (The StringIO 
workaround is also no longer needed)
